pkg/core/storage/memcached_store.go

```diff
@@ -83,21 +90,98 @@ func (s *MemCachedStore) GetBatch() *MemBatch {
 
 // Seek implements the Store interface.
 func (s *MemCachedStore) Seek(key []byte, f func(k, v []byte)) {
+	// Create memory store `mem` and `del` snapshots not to hold the lock.
+	memSnapshot := make([]KeyValue, 0)
+	delSnapshot := make(map[string]bool)
+	sk := string(key)
 	s.mut.RLock()
-	defer s.mut.RUnlock()
-	s.MemoryStore.seek(key, f)
-	s.ps.Seek(key, func(k, v []byte) {
-		elem := string(k)
-		// If it's in mem, we already called f() for it in MemoryStore.Seek().
-		_, present := s.mem[elem]
-		if !present {
-			// If it's in del, we shouldn't be calling f() anyway.
-			_, present = s.del[elem]
+	for k, v := range s.MemoryStore.mem {
+		if strings.HasPrefix(k, sk) {
+			memSnapshot = append(memSnapshot, KeyValue{
+				Key:   []byte(k),
+				Value: slice.Copy(v),
+			})
 		}
-		if !present {
-			f(k, v)
+	}
+	for k, v := range s.MemoryStore.del {
+		if strings.HasPrefix(k, sk) {
+			delSnapshot[k] = v
 		}
+	}
+	s.mut.RUnlock()
+	// Sort memSnapshot items for further comparison with ps items.
+	sort.Slice(memSnapshot, func(i, j int) bool {
+		return bytes.Compare(memSnapshot[i].Key, memSnapshot[j].Key) < 0
 	})
+
+	var (
+		data1   = make(chan KeyValue)
+		data2   = make(chan KeyValue)
+		seekres = make(chan KeyValue)
+	)
+	// Seek over memory store.
+	go func() {
+		for _, kv := range memSnapshot {
+			data1 <- kv
+		}
+		close(data1)
+	}()
+
+	// Seek over persistent store.
+	go func() {
+		n := len(memSnapshot)
+		s.mut.Lock()
+		s.ps.Seek(key, func(k, v []byte) {
+			// If it's in mem, we already called f() for it during memSnapshot iteration.
+			i := sort.Search(n, func(i int) bool {
+				return bytes.Compare(memSnapshot[i].Key, k) >= 0
+			})
+			if i < n && bytes.Equal(memSnapshot[i].Key, k) {
+				return
+			}
+			// If it's in del, we shouldn't be calling f() anyway.
+			if _, ok := delSnapshot[string(k)]; ok {
+				return
+			}
+			// Must copy here, #1468.
+			data2 <- KeyValue{
+				Key:   slice.Copy(k),
+				Value: slice.Copy(v),
+			}
+		})
+		s.mut.Unlock()
+		close(data2)
+	}()
+
+	// Merge results of seek operations in ascending order.
+	done := make(chan struct{})
+	go func() {
+		kvMem, haveMem := <-data1
+		kvPs, havePs := <-data2
+		for {
+			if !haveMem && !havePs {
+				break
+			}
+			var isMem = haveMem && (!havePs || (bytes.Compare(kvMem.Key, kvPs.Key) < 0))
```

Comment:
Here keys can't be equal, because we check the equality during seek over persistent storage:
```
if i < n && bytes.Equal(memSnapshot[i].Key, k) {
    return
}
```
So if persistent storage has the same key as memory storage, then the item from memory storage will be sent to the merging goroutine.

And we have a test for it, see `TestCachedSeekSorting`. There we have `updatedKVs` items - these items are in the persistent storage with `stub` value, and also in the memory storage with their real value - that's exactly the case when the same set of keys is in the memory and persistent storage.

Am I missing something?

Reply:
Yeah, I see now.